network/src/service/identify_service.rs

```diff
@@ -1,236 +1,151 @@
-            upgrade::apply(out.socket, IdentifyProtocolConfig, endpoint, client_addr).map(
-                move |(output, addr)| {
-                    let protocol = match output {
-                        IdentifyOutput::RemoteInfo {
-                            info,
-                            observed_addr,
-                        } => Protocol::IdentifyRequest(peer_id, info, observed_addr),
-                        IdentifyOutput::Sender { .. } => {
-                            panic!("should not reach here because we are dialer")
-                        }
-                    };
-                    (protocol, addr)
-                },
-            )
-        });
+        event_receiver: mpsc::UnboundedReceiver<IdentifyEvent>,
+    ) -> IdentifyService {
+        IdentifyService {
+            event_receiver,
+            network,
+            listen_addrs: HashMap::default(),
+        }
+    }
+}
 
-        let periodic_identify_future = Interval::new(
-            Instant::now() + Duration::from_secs(5),
-            self.identify_interval,
-        )
-        .map_err(|err| {
-            debug!(target: "network", "identify periodic error {:?}", err);
-            IoError::new(
-                IoErrorKind::Other,
-                format!("identify periodic error {:?}", err),
-            )
-        })
-        .for_each({
-            let transport = transport.clone();
-            let _identify_timeout = self.identify_timeout;
-            let network = Arc::clone(&network);
-            move |_| {
-                for peer_id in network.peers() {
-                    if let Some(ref identify_info) = network.get_peer_identify_info(&peer_id) {
-                        if identify_info.count_of_known_listen_addrs > 0 {
-                            continue;
-                        }
-                    }
-                    // TODO should we try all addresses?
-                    if let Some(addr) = network.get_peer_addresses(&peer_id).get(0) {
-                        trace!(
-                        target: "network",
-                        "request identify to peer {:?} {:?}",
-                        peer_id,
-                        addr
-                        );
-                        // dial identify
-                        let _ = swarm_controller.dial(addr.clone(), transport.clone());
-                    } else {
-                        error!(
-                        target: "network",
-                        "error when prepare identify : can't find addresses for peer {:?}",
-                        peer_id
-                        );
-                    }
+impl Stream for IdentifyService {
+    type Item = ();
+    type Error = ();
+    fn poll(&mut self) -> Result<Async<Option<Self::Item>>, Self::Error> {
+        match try_ready!(self.event_receiver.poll()) {
+            Some(IdentifyEvent::AddListenAddrs { peer_id, addrs }) => {
+                self.listen_addrs.insert(peer_id, addrs);
```

Comment:
should store addrs to peer_store if these addrs belongs to a peer.

Reply:
Refactor identify protocol later